### ISSUE TITLE
chained conditionals

### ISSUE THREAD

**Issue by maa105** [josdejong/mathjs#1262]
Hello, thanks for your efforts building and mainaining the library.

I noted that the following:
```
math.eval('50<T<100',{T:10})
```
returns true, while clearly it should return false.

any help is greatly appriciated

**Comment by ericman314** [comment]
Unfortunately, `50 < T < 100` doesn't work in math.js, or most programming languages for that matter, like you would expect. Instead, you have to do something like:

```js
 math.eval('50 < T and T < 100', {T:10})
```

**Comment by josdejong** [comment]
Indeed. It would be cool to have though, I really love this feature in python for example.

**Comment by ericman314** [comment]
I didn't know Python could do that. Looks like there are some other languages that can do it as well. We already implement complex parsing rules for things like conditional expressions (`expr ? a : b`), so maybe it wouldn't be too difficult to add support for chained comparison? I can't think of any reason why you would want `a < b < c` to be interpreted as `(a < b) < c`. Chaining the comparisons does make much more sense, and seems to be in line with math.js's goal of looking more "math-y".

**Comment by josdejong** [comment]
Agree!

Anyone interested in implementing chained comparison?

**Comment by josdejong** [comment]
Would be great if you could give it a try @ericman314

I think we should indeed allow any number of comparisons like `a < b < c < d < e`, and allow any of `<`, `<=`, `>`, and `>=` at any position. Some combinations don't make sense of course but I guess that's up to the user.

I think we could take two approaches: 

1. Implement a new node, like `RelationalNode`, which can hold a list with values and comparisons and evaluate it correctly.
2. Implement this as eye candy in the parser: parse `a < b < c` directly as `((a < b) and (b < c))` using three (existing) OperatorNodes.

What makes most sense to you?

**Comment by ericman314** [comment]
Python's implementation of chained comparison evaluates each operand only once and uses short-circuit evaluation. I think this would only be possible using a `RelationalNode` as you suggest in the first approach.

The second approach seems simpler but would evaluate some operands more than once and would not benefit from short-circuit evaluation.

I actually think the parser is designed well enough that the `RelationalNode` approach would not be too difficult, so I will give that a try first.

Should we also include `==` and `!=`? Those have equal precedence as the other comparison operators.

**Comment by josdejong** [comment]
I agree with you, `RelationalNode` feels like the best solution for me. Though it's more work, it will work much more logical when working with parsed node trees. For example the `toString()` of solution 2 will be very confusing.

> Should we also include == and !=? Those have equal precedence as the other comparison operators.

I'm not sure whether there is a usecase of using a combination of == and >, but it makes sense because then we handle all relational operators in an uniform way.

For backward compatibility I think it may make sense to return an `OperatorNode` when there is only one relational operation (current behavior), and return this new `RelationalNode` when there are two or more operators in play.

**Comment by ericman314** [comment]
> For backward compatibility I think it may make sense to return an OperatorNode when there is only one relational operation (current behavior), and return this new RelationalNode when there are two or more operators in play.

I was thinking the same thing.

> I'm not sure whether there is a usecase of using a combination of == and >

A possible pitfall is illustrated by this python program:

```py
if 3 < 4 == True:
  print "Expression 1"

if (3 < 4) == True:
  print "Expression 2"
```
Output:
```Expression 2```

In JavaScript, both `Expression 1` and `Expression 2` would be output, so the Python behavior feels weird to me. I'm OK either way, whether we want to include `==` and `!=`, or not.

**Comment by ericman314** [comment]
Got it working! Just have to do tests and other little things like `toLatex`.

**Comment by josdejong** [comment]
wow! that's fast! :rocket:

**Comment by josdejong** [comment]
This is available now in v5.2.0. Nice job Eric!
